In [8]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import random

# import ephys_unit_analysis as ena
import mz_ephys_unit_analysis as mz_ena
import mz_unit_dur_freq as mz_udf

%load_ext autoreload
%autoreload 2

In [ ]:
#V1_RSC_psth_G1GN = pd.read_hdf('u:/Papers/D&P Osc in Mouse Vis Sys/Data Analysis/Units/V1_RSC/V1_RSC_2kmeans_PSTH_GUI_mapped_sliced.hdf5')
#V1_RSC_spikes_G1GN = pd.read_hdf('u:/Papers/D&P Osc in Mouse Vis Sys/Data Analysis/Units/V1_RSC/V1_RSC_2kmeans_spikes_GUI_mapped_sliced.hdf5')


In [ ]:
# V1_LGN_spikes_G1GN = pd.read_hdf("U:\Papers\D&P Osc in Mouse Vis Sys\Data Analysis\Units\V1_LGN\V1_LGN_SPIKES_Final.hdf5")
# V1_LGN_psth_G1GN = pd.read_hdf("U:\Papers\D&P Osc in Mouse Vis Sys\Data Analysis\Units\V1_LGN\V1_LGN_PSTH_Final.hdf5")

In [ ]:
#master_psth_V1_LP_FINAL = pd.read_hdf('u:/Papers/D&P Osc in Mouse Vis Sys/Data Analysis/Units/V1_LP/LP_PSTH_remLT_300sps_remGT-0o10ampWF_LP_GUI.hdf5')
#master_spikes_V1_LP_FINAL = pd.read_hdf('u:/Papers/D&P Osc in Mouse Vis Sys/Data Analysis/Units/V1_LP/LP_spikes_remLT_300sps_remGT-0o10ampWF_LP_GUI.hdf5')


# load PSTH data

## V1-RSC, SK data

In [26]:
rsc_path = r"U:\Papers\DP Osc in Mouse Vis Sys\Data Analysis\Units\V1_RSC\V1_RSC_2kmeans_PSTH_GUI_mapped_sliced.hdf5"

store = pd.HDFStore(rsc_path)
store.keys()

['/master_psth']

In [27]:
v1rsc_df=store[store.keys()[0]]
v1rsc_df.head(2)

Hz              condition                  cuid  depth       et  \
150  4.368449  GroupWTpreG-1-G---GV1  001ET#R000id166preV1    800  ET#R000   
151  4.332749  GroupWTpreG-1-G---GV1  001ET#R000id166preV1    800  ET#R000   

     exp                                              fname    group location  \
150  001  u:/Data/skissing/V1_RSC/07.20.2019 V1_RSC_Pair...  GroupWT       V1   
151  001  u:/Data/skissing/V1_RSC/07.20.2019 V1_RSC_Pair...  GroupWT       V1   

      paradigm              rec              stim  times training    zscore  \
150  G-1-G---G  001ET#R000preV1  40TrialG-1-G---G   0.00      pre -0.152199   
151  G-1-G---G  001ET#R000preV1  40TrialG-1-G---G   0.01      pre -0.157560   

          ztc n_type layer  r_groups r_group_type  
150 -0.263173     rs  L2/3         1      excited  
151 -0.298374     rs  L2/3         1      excited

In [93]:
v1rsc_exc_df = v1rsc_df[v1rsc_df.r_group_type=='excited']

v1rsc_V1_units = v1rsc_exc_df[v1rsc_exc_df['condition']=='GroupWTpostG-1-G---GV1']
v1rsc_RSC_units = v1rsc_exc_df[v1rsc_exc_df['condition']=='GroupWTpostG-1-G---GRSC']

print(f'V1 units: {v1rsc_V1_units.cuid.nunique()} --- RSC units: {v1rsc_RSC_units.cuid.nunique()}')

V1 units: 318 --- RSC units: 218


## V1-HPC, MZ data

In [85]:
v1hpc_df = pd.read_pickle(r"C:\Users\AChub_Lab\Desktop\DIA_CQuinn\V1HPC_psth.pkl")
v1hpc_df.head()

times stim        Hz  cluster_id  depth    zscore       ztc     et  \
0   0.00  pre  1.236249         292   1060 -0.943946 -0.556308  et323   
1   0.01  pre  1.330091         292   1060 -0.874114 -0.460585  et323   
2   0.02  pre  1.433267         292   1060 -0.797336 -0.355341  et323   
3   0.03  pre  1.512557         292   1060 -0.738333 -0.274463  et323   
4   0.04  pre  1.522873         292   1060 -0.730656 -0.263940  et323   

         cc       cuid                                               path  \
0  cc033663  et323_292  G:\Neuropixels\Sam_multi_brain_region_paper\SO...   
1  cc033663  et323_292  G:\Neuropixels\Sam_multi_brain_region_paper\SO...   
2  cc033663  et323_292  G:\Neuropixels\Sam_multi_brain_region_paper\SO...   
3  cc033663  et323_292  G:\Neuropixels\Sam_multi_brain_region_paper\SO...   
4  cc033663  et323_292  G:\Neuropixels\Sam_multi_brain_region_paper\SO...   

  region  r_groups r_group_type  
0  hippo         1      excited  
1  hippo         1      excited  
2  hippo         1      excited  
3  hippo         1      excited  
4  hippo         1      excited

In [90]:
vr_v1hpc = v1hpc_df[v1hpc_df.r_group_type == 'excited']

v1hpc_V1_units = vr_v1hpc[vr_v1hpc['region']=='v1']
v1hpc_HPC_units = vr_v1hpc[vr_v1hpc['region']=='hippo']

for t,tt in vr_v1hpc.groupby(['region','stim']):
    print(f'{t} units: {tt.cuid.nunique()}')

('hippo', 'novel') units: 212
('hippo', 'post') units: 390
('hippo', 'pre') units: 245
('v1', 'novel') units: 416
('v1', 'post') units: 205
('v1', 'pre') units: 154


# Peak detection analysis - Avg number of units per mouse
- V1: 318 total, 216 w/ 2+, 163 w/ 3+
- RSC: 218 total, 62 w/ 2+, 30 w/ 3+
- V1: 339 total, ___ w/ 2+, 162 w/ 3+
- HPC: 324 total, ___ w/ 2+, 24 w/ 3+

## define functions

In [39]:
from detect_peaks import detect_peaks
def _duration_peaks_unit(unit):
    data = unit
    peakind = detect_peaks(data, mph=1.5, mpd=10, threshold=0)
    if peakind.size>0:
        if peakind.size>1:
            if peakind[0]>50 and peakind[0]<70:
                mask = np.array(np.diff(peakind)<50)
                for ind, val in enumerate(mask):
                    if val==False:
                        mask[ind:]=False
                mask = np.insert(mask, 0, True)
                peakind = peakind[mask]
                dur = (peakind[-1]/100) - 0.5
            else:   
                dur = float('nan')
        elif peakind[0]>50 and peakind[0]<70:
            dur = (peakind[0]/100) - 0.5
        else:
            dur = float('nan')
        if dur<0 or dur>2:
            dur = float('nan')
    else:
        dur = float('nan')
    return dur, peakind

In [40]:
def pull_osc_units(region_df, num_cycles, in_regions):
    num_peaks = num_cycles+1
    reg_df_temp = region_df.pivot(index='cuid', columns='times', values='zscore')
    reg_array = np.array(reg_df_temp)
    print(f'Pre peak detection: {region_df.cuid.nunique()} units')
    cuid, pk_ind, mouse = [], [], []
    for unit in range(reg_array.shape[0]):
        yy = reg_array[unit]
        dur, peak_ind = _duration_peaks_unit(yy)
        if len(peak_ind) >= num_peaks:
            cuid.append(reg_df_temp.index[unit])
            pk_ind.append(peak_ind)
            if in_regions == 'v1rsc':
                mouse.append(reg_df_temp.index[unit][3:10])
            elif in_regions == 'v1hpc':
                mouse.append(reg_df_temp.index[unit].split('_')[0])
    df = pd.DataFrame(zip(cuid,pk_ind,mouse), columns=['cuid','peak_ind','et'])
    print(f'Post peak detection: {df.cuid.nunique()} units')
    print(df.groupby('et')['cuid'].nunique())
    print(df.groupby('et')['cuid'].nunique().mean(), end='\n\n')
    return df

## V1-RSC oscillators

In [108]:
for i in [2,3]:
    print(f'Num cycles: {i}')
    v1rsc_V1_osc = pull_osc_units(v1rsc_V1_units, num_cycles=i, in_regions='v1rsc')

Num cycles: 2
Pre peak detection: 318 units
Post peak detection: 218 units
et
ET#R003    47
ET#R004    50
ET#RB00    48
ET#RB02    21
ET#RB03    52
Name: cuid, dtype: int64
43.6

Num cycles: 3
Pre peak detection: 318 units
Post peak detection: 108 units
et
ET#R003    24
ET#R004    28
ET#RB00    10
ET#RB02     8
ET#RB03    38
Name: cuid, dtype: int64
21.6



In [109]:
for i in [2,3]:
    print(f'Num cycles: {i}')
    v1rsc_RSC_osc = pull_osc_units(v1rsc_RSC_units, num_cycles=i, in_regions='v1rsc')

Num cycles: 2
Pre peak detection: 218 units
Post peak detection: 99 units
et
ET#R003    14
ET#R004     4
ET#RB00    25
ET#RB02    11
ET#RB03    45
Name: cuid, dtype: int64
19.8

Num cycles: 3
Pre peak detection: 218 units
Post peak detection: 46 units
et
ET#R003     6
ET#R004     1
ET#RB00    12
ET#RB02     4
ET#RB03    23
Name: cuid, dtype: int64
9.2



## V1-HPC

In [127]:
for i in [2,3]:
    print(f'Num cycles: {i}')
    v1hpc_V1_osc = pull_osc_units(v1hpc_V1_units, num_cycles=i, in_regions='v1hpc')

Num cycles: 2
Pre peak detection: 372 units
Post peak detection: 176 units
et
et1        2
et10       5
et2       14
et20      17
et200      5
et3       17
et30       5
et311     15
et3112    15
et323     20
et3232    20
et383     18
et3832    23
Name: cuid, dtype: int64
13.538461538461538

Num cycles: 3
Pre peak detection: 372 units
Post peak detection: 57 units
et
et10       1
et2        7
et20      10
et200      3
et3        8
et30       1
et311      2
et3112     1
et323      3
et3232     5
et383      7
et3832     9
Name: cuid, dtype: int64
4.75



In [129]:
for i in [2,3]:
    print(f'Num cycles: {i}')
    v1hpc_HPC_osc = pull_osc_units(v1hpc_HPC_units, num_cycles=i, in_regions='v1hpc')

Num cycles: 2
Pre peak detection: 475 units
Post peak detection: 181 units
et
et1       12
et10      15
et2        8
et20      14
et200      7
et3       17
et30      10
et311     21
et3112     9
et323     12
et3232    13
et383     18
et3832    25
Name: cuid, dtype: int64
13.923076923076923

Num cycles: 3
Pre peak detection: 475 units
Post peak detection: 48 units
et
et10      8
et2       1
et20      3
et200     1
et3       6
et30      4
et311     8
et3112    1
et323     1
et3232    5
et383     5
et3832    5
Name: cuid, dtype: int64
4.0

